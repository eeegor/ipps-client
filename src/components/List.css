.ReactVirtualized__AutoSizer {
  width: 100% !important;
  height: 100vh !important;
}
.ReactVirtualized__Grid.ReactVirtualized__Table__Grid {
  height: calc(100vh - 40px) !important;
  overflow: scroll !important;
}

.ReactVirtualized__Table {
  height: 100%;
  font-size: 0.8rem;
  overflow: scroll !important;
}

.ReactVirtualized__Table__row,
.ReactVirtualized__Table__headerRow {
  display: grid;
  grid-template-columns: 600px 300px 200px 200px 300px 300px 300px;
  text-align: left;
  background: #fff;
  white-space: nowrap;
  overflow: hidden !important;
}

.ReactVirtualized__Table__headerRow {
  overflow-x: scroll !important;
  overflow-y: visible !important;
  align-items: center;
  border-bottom: 1px solid #ddd;
  background: #fafafa;
  z-index: 10000;
}

.ReactVirtualized__Table__row {
  border-bottom: 1px solid #ccc;
}

.ReactVirtualized__Table__headerColumn {
  display: flex;
  padding: 0.2rem 1rem 0.1rem 1rem;
  align-items: center;
  border-right: 1px solid #ddd;
  font-weight: 700;
}

.ReactVirtualized__Table__rowColumn {
  padding: 0.55rem 1rem 0.45rem 1rem;
  margin-bottom: -1px;
  display: flex;
  align-items: center;
  border-right: 1px solid #f5f5f5;
}
