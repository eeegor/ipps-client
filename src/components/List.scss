@import '../variables';

.ReactVirtualized__AutoSizer {
  width: 100% !important;
  height: calc(100vh - 4rem) !important;
}
.ReactVirtualized__Grid.ReactVirtualized__Table__Grid {
  height: calc(100vh - 4rem - 40px) !important;
  overflow: scroll !important;
  -webkit-overflow-scrolling: touch;
}
@media (min-width: $size--tablet) {
  .ReactVirtualized__AutoSizer {
    height: calc(100vh - 5rem) !important;
  }
  .ReactVirtualized__Grid.ReactVirtualized__Table__Grid {
    height: calc(100vh - 5rem - 40px) !important;
  }
}

.ReactVirtualized__Table {
  height: 100%;
  font-size: 0.8rem;
  overflow: scroll !important;
  -webkit-overflow-scrolling: touch;
}

.ReactVirtualized__Table__row,
.ReactVirtualized__Table__headerRow {
  display: grid;
  grid-template-columns: 500px 400px 200px 160px 360px 180px 250px 250px 300px;
  text-align: left;
  background: #fff;
  white-space: nowrap;
  overflow: hidden !important;
}

.ReactVirtualized__Table__headerRow {
  overflow-x: hidden !important;
  overflow-y: visible !important;
  align-items: center;
  border-bottom: 1px solid #ddd;
  background: #fff;
  z-index: 10000;
}

.ReactVirtualized__Table__row {
  border-bottom: 1px solid #ccc;

  &:hover {
    background: rgba($color--primary, 0.05);
  }
}

.ReactVirtualized__Table__headerColumn {
  display: flex;
  padding: 0.2rem 1rem 0.1rem 1rem;
  align-items: center;
  border-right: 1px solid #ddd;
  font-weight: 700;
}

.ReactVirtualized__Table__rowColumn {
  padding: 0.55rem 1rem 0.45rem 1rem;
  margin-bottom: -1px;
  display: flex;
  align-items: center;
  border-right: 1px solid #f5f5f5;
}

.ReactVirtualized__Table__sortableHeaderIcon {
  fill: $color--primary;
}
