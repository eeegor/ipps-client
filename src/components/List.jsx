import React from 'react';
import { AutoSizer, Table, Column, SortDirection } from 'react-virtualized';
import { sortBy as lodashSortBy } from 'lodash';
import './List.scss';

export class List extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      sortBy: this.props.columns[0].dataKey || this.props.columns[0].label,
      sortDirection: SortDirection.DESC,
      sortedItems: lodashSortBy(
        this.props.items,
        this.props.columns[0].dataKey || this.props.columns[0].label
      )
    };

    this._sort = this._sort.bind(this);
  }

  componentDidUpdate(prevProps) {
    const prev = prevProps.items || null;
    const next = this.props.items || null;
    if (prev !== next) {
      this.setState({
        sortedItems: lodashSortBy(this.props.items, this.state.sortBy)
      });
    }
  }

  _sort(sortProps) {
    const { sortBy, sortDirection } = sortProps;
    const sortedItems = this._getSortedList(
      this.state.sortedItems,
      sortBy,
      sortDirection
    );

    if (sortedItems.length > 0) {
      this.setState({
        sortBy,
        sortDirection,
        sortedItems
      });
    }
  }

  _getSortedList(items, sortBy, sortDirection) {
    const key = sortBy;
    const currencyKeys = [
      'Average Covered Charges',
      'Average Total Payments',
      'Average Medicare Payments'
    ];
    const sorted = lodashSortBy(items, item =>
      currencyKeys.includes(key)
        ? parseFloat(item[key].replace(/[$|,]/g, ''))
        : item[key]
    );
    return sortDirection === SortDirection.ASC ? sorted : sorted.reverse();
  }

  render() {
    const { sortBy, sortDirection, sortedItems } = this.state;
    const { columns, windowSize } = this.props;
    const cols = columns.map((columnProps, index) => (
      <Column
        key={index}
        width={1}
        dataKey={columnProps['dataKey'] || columnProps.label}
        {...columnProps}
      />
    ));

    const calculatedHeight = windowSize && {
      height: windowSize.height - (windowSize.width < 767 ? 64 : 80)
    };

    return (
      <AutoSizer
        className="ReactVirtualized__AutoSizer"
        style={calculatedHeight}
      >
        {() => (
          <Table
            headerHeight={40}
            rowCount={sortedItems.length}
            rowGetter={({ index }) => sortedItems[index]}
            rowHeight={40}
            width={2600}
            height={calculatedHeight.height}
            sort={this._sort}
            sortBy={sortBy}
            sortDirection={sortDirection}
          >
            {cols}
          </Table>
        )}
      </AutoSizer>
    );
  }
}
