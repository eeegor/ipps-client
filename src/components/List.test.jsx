import React from 'react';
import renderer from 'react-test-renderer';
import { mount, shallow } from 'enzyme';
import { List } from './List';

let wrapper;

beforeEach(() => {
  const columns = [{ dataKey: 'a', label: 'A' }, { dataKey: 'b', label: 'B' }];
  wrapper = mount(<List columns={columns}>Hello</List>);
});

describe('List', () => {
  it('renders markup correctly', () => {
    const tree = renderer.create(wrapper).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
