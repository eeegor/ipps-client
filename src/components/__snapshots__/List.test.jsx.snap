// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List renders markup correctly 1`] = `
<div
  className="ReactVirtualized__AutoSizer"
  style={
    Object {
      "height": 0,
      "overflow": "visible",
      "width": 0,
    }
  }
>
  <div
    aria-colcount={2}
    aria-rowcount={0}
    className="ReactVirtualized__Table"
    role="grid"
    style={Object {}}
  >
    <div
      className="ReactVirtualized__Table__headerRow"
      role="row"
      style={
        Object {
          "height": 40,
          "overflow": "hidden",
          "paddingRight": 0,
          "width": 2600,
        }
      }
    >
      <div
        aria-label="A"
        aria-sort="descending"
        className="ReactVirtualized__Table__headerColumn ReactVirtualized__Table__sortableHeaderColumn"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="columnheader"
        style={
          Object {
            "WebkitFlex": "0 1 1px",
            "flex": "0 1 1px",
            "msFlex": "0 1 1px",
          }
        }
        tabIndex={0}
      >
        <span
          className="ReactVirtualized__Table__headerTruncatedText"
          title="A"
        >
          A
        </span>
        <svg
          className="ReactVirtualized__Table__sortableHeaderIcon ReactVirtualized__Table__sortableHeaderIcon--DESC"
          height={18}
          viewBox="0 0 24 24"
          width={18}
        >
          <path
            d="M7 10l5 5 5-5z"
          />
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
        </svg>
      </div>
      <div
        aria-label="B"
        aria-sort="none"
        className="ReactVirtualized__Table__headerColumn ReactVirtualized__Table__sortableHeaderColumn"
        onClick={[Function]}
        onKeyDown={[Function]}
        role="columnheader"
        style={
          Object {
            "WebkitFlex": "0 1 1px",
            "flex": "0 1 1px",
            "msFlex": "0 1 1px",
          }
        }
        tabIndex={0}
      >
        <span
          className="ReactVirtualized__Table__headerTruncatedText"
          title="B"
        >
          B
        </span>
      </div>
    </div>
    <div
      aria-label="grid"
      aria-readonly={true}
      className="ReactVirtualized__Grid ReactVirtualized__Table__Grid"
      onScroll={[Function]}
      role="rowgroup"
      style={
        Object {
          "WebkitOverflowScrolling": "touch",
          "boxSizing": "border-box",
          "direction": "ltr",
          "height": -40,
          "overflowX": "hidden",
          "overflowY": "auto",
          "position": "relative",
          "width": 2600,
          "willChange": "transform",
        }
      }
      tabIndex={0}
    />
  </div>
</div>
`;
